powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/behavior/tasks/v3/ActivationServiceBehavior.java

```diff
@@ -1105,63 +1142,229 @@ public CreateActivationResponse createActivation(
+        final ActivationOtpValidation expectedStage = activationOtpValidationConverter.convertFrom(activation.getActivationOtpValidation());
+        final String expectedOtpHash = activation.getActivationOtp();
+
+        if (currentStage == ActivationOtpValidation.NONE) {
+            // This should never happen.
+            logger.info("Internal error in activation OTP validation: {}", activationId);
+            // Rollback is not required, database is not used for writing yet.
+            throw localizationProvider.buildExceptionForCode(ServiceError.UNKNOWN_ERROR);
+        }
+
+        // Check whether activation OTP validation is turned OFF. In this case, the confirmation OTP must not
+        // be provided.
+        if (expectedStage == ActivationOtpValidation.NONE) {
+            if (confirmationOtp != null) {
+                logger.info("Activation OTP is not used, but is provided: {}", activationId);
+                // Rollback is not required, database is not used for writing yet.
+                throw localizationProvider.buildExceptionForCode(ServiceError.INVALID_ACTIVATION_OTP);
+            }
+            return;
+        }
+        // Check whether this is validation in the different step of activation. If yes, then the confirmation
+        // OTP must not be provided.
+        if (expectedStage != currentStage) {
+            if (confirmationOtp != null) {
+                logger.info("Activation OTP is not expected, but is provided: {}", activationId);
+                // Rollback is not required, database is not used for writing yet.
+                throw localizationProvider.buildExceptionForCode(ServiceError.INVALID_ACTIVATION_OTP);
+            }
+            return;
+        }
+        // We're in the right step, so the confirmation OTP must be provided.
+        if (confirmationOtp == null) {
+            logger.info("Activation OTP is expected, but is missing: {}", activationId);
+            // Rollback is not required, database is not used for writing yet.
+            throw localizationProvider.buildExceptionForCode(ServiceError.INVALID_ACTIVATION_OTP);
+        }
+        // The final test only checks, whether hash is present in the database.
+        if (expectedOtpHash == null) {
+            logger.info("Activation OTP is missing in activation data: {}", activationId);
+            // Rollback is not required, database is not used for writing yet.
+            throw localizationProvider.buildExceptionForCode(ServiceError.INVALID_ACTIVATION_OTP);
+        }
+
+        // Now verify OTP value
+        try {
+            if (PasswordHash.verify(confirmationOtp.getBytes(StandardCharsets.UTF_8), expectedOtpHash)) {
+                // Everything looks fine. Reset the failed attempts counter.
+                activation.setFailedAttempts(0L);
+                return;
+            }
+        } catch (IOException e) {
+            logger.warn("Invalid activation OTP hash: {}", activationId);
```

Comment:
Can we add a comment that this error does not cause an exception and requires execution of the rest of the method? It is not visible on first sight.